consensus/src/marshal/actor.rs

```diff
@@ -0,0 +1,764 @@
+            Ok(_) => {
+                debug!(view, "verified stored");
+            }
+            Err(archive::Error::AlreadyPrunedTo(_)) => {
+                debug!(view, "verified already pruned");
+            }
+            Err(e) => {
+                panic!("Failed to insert verified block: {e}");
+            }
+        }
+    }
+
+    /// Add a notarization (with block) to the archive.
+    async fn put_notarization(
+        &mut self,
+        view: u64,
+        commitment: B::Commitment,
+        notarization: Notarization<V, B::Commitment>,
+        block: B,
+    ) {
+        match self
+            .notarized
+            .put_sync(view, commitment, (notarization, block))
+            .await
+        {
+            Ok(_) => {
+                debug!(view, "notarized stored");
+            }
+            Err(archive::Error::AlreadyPrunedTo(_)) => {
+                debug!(view, "notarized already pruned");
+            }
+            Err(e) => {
+                panic!("Failed to insert notarization: {e}");
+            }
+        }
+    }
+
+    /// Add a (finalized) block to the archive.
+    ///
+    /// At the end of the method, the notifier is notified to indicate that there has been an update
+    /// to the archive of (finalized) blocks.
+    async fn put_block(
+        &mut self,
+        height: u64,
+        commitment: B::Commitment,
+        block: B,
+        notifier: &mut mpsc::Sender<()>,
+    ) {
+        if let Err(e) = self.blocks.put_sync(height, commitment, block).await {
+            panic!("Failed to insert block: {e}");
+        }
+        let _ = notifier.try_send(());
+    }
+
+    /// Add a finalization (with block) to the archive.
+    async fn put_finalization(
+        &mut self,
+        height: u64,
+        commitment: B::Commitment,
+        finalization: Finalization<V, B::Commitment>,
+        block: B,
+        notifier: &mut mpsc::Sender<()>,
+    ) {
+        if let Err(e) = self
+            .finalized
+            .put_sync(height, commitment, finalization)
+            .await
+        {
+            panic!("Failed to insert finalization: {e}");
+        }
+        self.put_block(height, commitment, block, notifier).await;
+    }
+
+    /// Looks for a block in local storage.
+    ///
+    /// Tries to find the block efficiently, starting with the buffer, then verified,
+    /// then notarized, and finally the blocks archive.
+    async fn search_for_block(
+        &mut self,
+        buffer: &mut buffered::Mailbox<P, B>,
+        commitment: B::Commitment,
+        depth: SearchDepth,
+    ) -> Option<B> {
+        // Check buffer
+        if let Some(block) = buffer.get(None, commitment, None).await.into_iter().next() {
+            return Some(block);
+        }
+
+        // Check verified
+        if let Some(block) = self.get_verified(Identifier::Key(&commitment)).await {
+            return Some(block);
+        }
+        if depth < SearchDepth::Notarized {
```

Comment:
Cool abstraction 👀

Reply:
Might be unnecessary complexity, I might instead just do something like:

```
async fn search_a_b_c(...) {
  self.search_a_b(...).await;
  self.c.find(...);
}
async fn search_a_b(...) {
  self.search_a(...).await;
  self.b.find(...);
}
async fn search_a(...) {
  self.a.find(...);
}
```